pkg/controller/apiextensions/definer/api.go

```diff
@@ -0,0 +1,109 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package definer
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metaapi "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1/instance"
+)
+
+const (
+	errDeleteNonControlledCRD = "cannot delete a crd that the definer does not control"
+	errGenerateCRD            = "cannot generate crd for given infrastructure definition"
+)
+
+// NewAPIDefaultCRDManager returns a new APIDefaultCRDManager.
+func NewAPIDefaultCRDManager(client client.Client) CRDManager {
+	return &APIDefaultCRDManager{client: resource.ClientApplicator{
+		Client:     client,
+		Applicator: resource.NewAPIPatchingApplicator(client),
+	}}
+}
+
+// APIDefaultCRDManager manages the generated CRD.
+type APIDefaultCRDManager struct {
+	client resource.ClientApplicator
+}
+
+// Get fetches the CRD.
+func (m *APIDefaultCRDManager) Get(ctx context.Context, definer CRDDefiner) (*v1beta1.CustomResourceDefinition, error) {
+	crd := &v1beta1.CustomResourceDefinition{}
+	return crd, m.client.Get(ctx, types.NamespacedName{Name: definer.GetCRDName()}, crd)
+}
+
+// Apply applies the CRD that the definer generates.
+func (m *APIDefaultCRDManager) Apply(ctx context.Context, definer CRDDefiner) error {
+	generated, err := definer.GenerateCRD()
+	if err != nil {
+		return errors.Wrap(err, errGenerateCRD)
+	}
+	meta.AddOwnerReference(generated, meta.AsController(meta.ReferenceTo(definer, definer.GetObjectKind().GroupVersionKind())))
+	return m.client.Apply(ctx, generated, resource.MustBeControllableBy(definer.GetUID()))
+}
+
+// Delete the generated CRD.
+func (m *APIDefaultCRDManager) Delete(ctx context.Context, definer CRDDefiner) error {
+	crd := &v1beta1.CustomResourceDefinition{}
+	err := m.client.Get(ctx, types.NamespacedName{Name: definer.GetCRDName()}, crd)
+	if resource.IgnoreNotFound(err) != nil {
+		return err
+	}
+	if kerrors.IsNotFound(err) {
+		return nil
+	}
+	if !metav1.IsControlledBy(crd, definer) {
+		return errors.New(errDeleteNonControlledCRD)
+	}
+	return resource.IgnoreNotFound(m.client.Delete(ctx, crd))
+}
+
+// DeleteInstances deletes all instances of the generated CRD in all namespaces
+// and returns whether there are any remaining instances.
+func (m *APIDefaultCRDManager) DeleteInstances(ctx context.Context, definer CRDDefiner) (bool, error) {
```

Comment:
Nit: `DeleteResources`? An "instance of a crd" is the thing defined by the resource definition, i.e. a resource.

Reply:
Also, this is the same behaviour as deleting a CRD, right? i.e. When you delete a CRD all of the custom resources it defined are deleted?

Reply:
> An "instance of a crd" is the thing defined by the resource definition, i.e. a resource.

This is correct but I think `resource` is kind of an overloaded term in the codebase and refers to `a Kubernetes resource` rather than `a Kubernetes Custom Resource`. _Deleting instances of a definer_ sounds more explicit to me.

> Also, this is the same behaviour as deleting a CRD, right? i.e. When you delete a CRD all of the custom resources it defined are deleted?

Yes, we have to delete the instances manually. It's explained in the function as:
> // NOTE(muvaf): When user deletes InfrastructureDefinition object the deletion
	// signal does not cascade to the owned resource until owner is gone. But
	// owner has its own finalizer that depends on having no instance of the CRD
	// because it cannot go away before stopping the controller.
	// So, we need to delete all instances of CRD manually here.